qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java

```diff
@@ -186,15 +199,79 @@ private static void doCodeActionToDisableValidation(Template template, List<Diag
 		codeActions.add(disableValidationForTemplateQuickFix);
 	}
 
+	/**
+	 * Create CodeAction for deprecated `with` Qute syntax.
+	 *
+	 * @param template    the Qute template.
+	 * @param diagnostic  the diagnostic list that this CodeAction will fix.
+	 * @param codeActions the list of CodeActions to perform.
+	 * @throws BadLocationException
+	 *
+	 */
+	private void doCodeActionsForNotRecommendedWithSection(Template template, Diagnostic diagnostic,
+			List<CodeAction> codeActions) {
+		Range withSectionRange = diagnostic.getRange();
+		try {
+			int withSectionStart = template.offsetAt(withSectionRange.getStart());
+			WithSection withSection = (WithSection) template.findNodeAt(withSectionStart + 1);
+			String javaResourceClassName = withSection.getObjectParameter().getName();
+			List<Node> withExpressions = findExpressions(withSection);
+			List<String> letExpressionParameterList = new ArrayList<String>();
+			for (Node expression : withExpressions) {
+				letExpressionParameterList.add(MessageFormat.format("{1}={0}.{1}", javaResourceClassName,
+						((Expression) expression).getContent()));
+			}
+			String letExpressionParameters = String.join(" ", letExpressionParameterList);
+
+			List<TextEdit> edits = new ArrayList<TextEdit>();
+
+			String withSectionOpenReplacementText = MessageFormat.format("#let {0}", letExpressionParameters);

```

Comment:
I find it's difficult to read the code, because:

 * there are no comments which explain each text edit which are generated
 * code is a litlle mixed (ex : letExpressionParameters is computed, you declare List<TextEdit> edits,and after you create the text edit by using letExpressionParameters ). 

To fix that,I suggest that crate  static methods createWithSectionOpenEdit, createInsertLetParametersEdit, createWithSectionCloseEdit, your organize your code like this:

```java
// 1. Get the #with section node
WithSection withSection = (WithSection) template.findNodeAt(withSectionStart + 1);

// 2. Create required text edits to update #with to #let
List<TextEdit> edits = new ArrayList<TextEdit>();

// 2.1 Create text edit which update start section #with to #let
TextEdit withSectionOpenEdit= createWithSectionOpenEdit(withSection);
edits.add(withSectionOpenEdit);

// 2.2 Define all expression / parameter section in the #let section as parameters
TextEdit insertLetParametersEdit = createInsertLetParametersEdit(withSection);
if (insertLetParametersEdit != null) {
  edits.add(updateSectionStartEdit );
}

// 2.3 Create text edit which updates end section \with to \let
TextEdit withSectionCloseEdit = createWithSectionCloseEdit(withSection);
// Take care of null, because of #with is end with {\} you must not generate a text edit 
// Please write test with that
if (updateSectionEndEdit  != null) {
  edits.add(updateSectionStartEdit );
}

CodeAction replaceWithSection = ...
...
```